pointmatcher/IO.cpp

```diff
@@ -1199,28 +1218,99 @@ typename PointMatcher<T>::DataPoints PointMatcherIO<T>::loadVTK(std::istream& is
-						throw runtime_error(string("Field " + fieldName + " is " + type + " but can only be of type double or float"));
+			Matrix descriptorData(dim, pointCount);
 
+			
+			// Load time data
+			if(isTimeSec || isTimeNsec)
+			{
 				// Skip LOOKUP_TABLE line
 				if(skipLookupTable)
 				{
 					getline(is, line);
 				}
-				readVtkData(type, isBinary, descriptor.transpose(), is);
+				
+				typename std::map<std::string, SplitTime>::iterator it;
+
+				it = labelledSplitTime.find(name);
+				// reserve memory
+				if(it == labelledSplitTime.end())
+				{
+					SplitTime t;
+					t.sec = Eigen::Matrix<unsigned int, Eigen::Dynamic, Eigen::Dynamic> (dim, pointCount);
+					t.nsec = t.sec;
+					labelledSplitTime[name] = t;
+				}
+
+				// Load seconds
+				if(isTimeSec)
+				{
+					assert(labelledSplitTime[name].isSecFound == false);
+					readVtkData(type, isBinary, labelledSplitTime[name].sec.transpose(), is);
+					labelledSplitTime[name].isSecFound = true;
+				}
+				
+				// Load nano seconds
+				if(isTimeNsec)
+				{
+					assert(labelledSplitTime[name].isNsecFound == false);
+					readVtkData(type, isBinary, labelledSplitTime[name].nsec.transpose(), is);
+					labelledSplitTime[name].isNsecFound = true;
+				}
 			}
-			loadedPoints.addDescriptor(name, descriptor);
+			else
+			{
+				if(isColorScalars && isBinary) 
+				{
+					std::vector<unsigned char> buffer(dim);
+					for (int i = 0; i < pointCount; ++i){
+						is.read(reinterpret_cast<char *>(&buffer.front()), dim);
+						for(int r=0; r < dim; ++r){
+							descriptorData(r, i) = buffer[r] / static_cast<T>(255.0);
+						}
+					}
+				} 
+				else 
+				{
+					if(!(type == "float" || type == "double" || type == "unsigned_int"))
+						throw runtime_error(string("Field " + fieldName + " is " + type + " but can only be of type double, float or unsigned_int."));
+
+					// Skip LOOKUP_TABLE line
+					if(skipLookupTable)
+					{
+						getline(is, line);
+					}
+					readVtkData(type, isBinary, descriptorData.transpose(), is);
+				}
+				loadedPoints.addDescriptor(name, descriptorData);
+			}
+		}
+	}
+
+	// Combine time and add to point cloud
+	typename std::map<std::string, SplitTime>::iterator it;
+	for(it=labelledSplitTime.begin(); it!=labelledSplitTime.end(); it++)
+	{
+		// Confirm that both parts were loaded
+		if(it->second.isSecFound == false)
+		{
+			throw runtime_error(string("Missing time field representing seconds. Expecting SCALARS with name " + it->first + "_splitTime_sec in the VTK file."));
+		}
+		
+		if(it->second.isNsecFound == false)
+		{
+			throw runtime_error(string("Missing time field representing nano seconds. Expecting SCALARS with name " + it->first + "_splitTime_nsec in the VTK file."));
 		}
+
+		// Loop through points
+		Int64Matrix timeData(1,pointCount);
+		for(int i=0; i<it->second.sec.cols(); i++)
+		{
+		
+			timeData(0,i) = (((boost::int64_t) it->second.sec(0,i)) << 32) | ((boost::int64_t) it->second.nsec(0,i));
```

Comment:
Assuming the int64 value should be the signed nanoseconds since the epoch higher 32 bit word will not correspond to seconds since epoch in almost any case, simply because e.g. 1 in the upper word will yield a 64bit value of 2^32 and that is quite far away from 10^9, which is the number of nanoseconds in a second. And that makes the identifier `sec` very misleading - I would call it wrong.